src/NLog/Internal/FileAppenders/BaseFileAppenderCache.cs

```diff
@@ -0,0 +1,258 @@
+                {
+                    // found it, move it to the first place on the list
+                    // (MRU)
+
+                    // file open has a chance of failure
+                    // if it fails in the constructor, we won't modify any data structures
+                    BaseFileAppender app = appenders[i];
+                    for (int j = i; j > 0; --j)
+                    {
+                        appenders[j] = appenders[j - 1];
+                    }
+
+                    appenders[0] = app;
+                    appenderToWrite = app;
+                    break;
+                }
+            }
+
+            if (appenderToWrite == null)
+            {
+                BaseFileAppender newAppender = Factory.Open(fileName, CreateFileParameters);
+
+                if (appenders[freeSpot] != null)
+                {
+                    appenders[freeSpot].Close();
+                    appenders[freeSpot] = null;
+                }
+
+                for (int j = freeSpot; j > 0; --j)
+                {
+                    appenders[j] = appenders[j - 1];
+                }
+
+                appenders[0] = newAppender;
+                appenderToWrite = newAppender;
+            }
+
+            return appenderToWrite;
+        }
+
+        public void CloseAppenders()
+        {
+            if (appenders != null)
+            {
+                for (int i = 0; i < appenders.Length; ++i)
+                {
+                    if (appenders[i] == null)
+                    {
+                        break;
+                    }
+
+                    appenders[i].Close();
+                    appenders[i] = null;
+                }
+            }
+        }
+
+        public void CloseAppenders(DateTime expireTime)
+        {
+            for (int i = 0; i < this.appenders.Length; ++i)
+            {
+                if (this.appenders[i] == null)
+                {
+                    break;
+                }
+
+                if (this.appenders[i].OpenTime < expireTime)
+                {
+                    for (int j = i; j < this.appenders.Length; ++j)
+                    {
+                        if (this.appenders[j] == null)
+                        {
+                            break;
+                        }
+
+                        this.appenders[j].Close();
+                        this.appenders[j] = null;
+                    }
+
+                    break;
+                }
+            }
+        }
+
+        // HACK: This method exposes a reference to the underlying item directly to the calling method. 
+        //    A new object should be constructed through a copy constructor. At the moment we will allow this until we
+        //    verify that there is not significant performance penalty.
+        public BaseFileAppender FindItem(string fileName)
+        {
+            foreach (BaseFileAppender appender in appenders)
+            {
+                if (appender == null)
```

Comment:
can we remove this duplicate code with an helper, which using `yield`? I propose: getAppenders